envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/LuaFilterFactory.kt

```diff
@@ -41,78 +44,103 @@ class LuaFilterFactory(private val incomingPermissionsProperties: IncomingPermis
-                                    incomingPermissionsProperties.requestIdentificationHeaders
-                                        .map { Value.newBuilder().setStringValue(it).build() }
-                                )
-                                .build()
-                            ).build()
-                    )
-                    .putFields(
-                        "trusted_client_identity_header",
-                        Value.newBuilder()
-                            .setStringValue(trustedClientIdentityHeader)
-                            .build()
-                    )
-                    .putFields(
-                        "san_uri_lua_pattern",
-                        Value.newBuilder()
-                            .setStringValue(sanUriWildcardRegexForLua).build()
-                    ).putFields("clients_allowed_to_all_endpoints",
-                        Value.newBuilder()
-                            .setListValue(ListValue.newBuilder()
-                                .addAllValues(
-                                    incomingPermissionsProperties.clientsAllowedToAllEndpoints
-                                        .map { Value.newBuilder().setStringValue(it).build() }
-                                )
-                                .build()
-                            ).build()
-                    )
-                    .putFields(
-                        "service_name",
-                        Value.newBuilder()
-                            .setStringValue(group.serviceName)
-                            .build()
-                    )
-                    .putFields(
-                        "discovery_service_name",
-                        Value.newBuilder()
-                            .setStringValue(group.discoveryServiceName ?: "")
-                            .build()
-                    ).putFields("flags", Value.newBuilder()
-                        .setStructValue(Struct.newBuilder()
-                            .apply {
-                                flags.forEach {
-                                    putFields(
-                                        it.key,
-                                        Value.newBuilder().setBoolValue(it.value).build()
-                                    )
-                                }
-                            }
-                            .build()
-                        ).build()
-                    )
-                    .build()
-            ).build()
+            .putFilterMetadata("envoy.filters.http.lua", metadata.toValue().structValue)
+            .build()
     }
 
     fun ingressClientNameHeaderFilter(): HttpFilter? =
         ingressClientNameHeaderFilter.takeIf { trustedClientIdentityHeader.isNotEmpty() }
 }
+
+sealed class LuaMetadataProperty<T>(open val value: T) {
+    abstract fun toValue(): Value
+
+    data class BooleanPropertyLua(override val value: Boolean) : LuaMetadataProperty<Boolean>(value) {
```

Comment:
Correct me if I misunderstand. We create `n*m` this objects, where `n` is number of groups and `m` is number of boolean properites? If so, I think it would be beter to make constructor private and instead create factory method to return this objects. This allow us to create static variables for them and reduced number of this objects to 2. Then `Value` object can be static to, which reduced creation of objects.